suite: add option to compile zlib

woot000 opened the pull request:
Adds the option to compile zlib along with a selection of forks compatible with zlib's API (zlib-chromium, zlib-cloudflare, and zlib-ng).

I didn't take the approach of installing zlib-ng as an msys2 package [like I had suggested a few months ago](https://github.com/m-ab-s/media-autobuild_suite/issues/2812#issue-2760276465) because I didn't feel like waiting for it to be added much longer. I also tried adding zlib-rs too, but after building (what I think was) a valid zlib-API compatible library I got this error trying to link it with ffmpeg (I may try adding zlib-rs at a later date, so this is just here for my own reference):
```
ccache gcc -D_ISOC11_SOURCE -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -DWIN32_LEAN_AND_MEAN -U__STRICT_ANSI__ -D__USE_MINGW_ANSI_STDIO=1 -D__printf__=__gnu_printf__ -D_POSIX_C_SOURCE=200112 -D_XOPEN_SOURCE=600 -DPIC -D_FORTIFY_SOURCE=2 -fstack-protector-strong -pipe -D__USE_MINGW_ANSI_STDIO=1 -mthreads -Wno-int-conversion -std=c17 -fomit-frame-pointer -g -Wdeclaration-after-statement -Wall -Wdisabled-optimization -Wpointer-arith -Wredundant-decls -Wwrite-strings -Wtype-limits -Wundef -Wmissing-prototypes -Wstrict-prototypes -Wempty-body -Wno-parentheses -Wno-switch -Wno-format-zero-length -Wno-pointer-sign -Wno-unused-const-variable -Wno-bool-operation -Wno-char-subscripts -O3 -fno-math-errno -fno-signed-zeros -fno-tree-vectorize -Werror=format-security -Werror=implicit-function-declaration -Werror=missing-prototypes -Werror=return-type -Werror=vla -Wformat -fdiagnostics-color=auto -Wmaybe-uninitialized -c -o /tmp/ffconf.Y7bFUbHS/test.o /tmp/ffconf.Y7bFUbHS/test.c
ar rcD /tmp/ffconf.Y7bFUbHS/test.a @/dev/null
check_builtin stdbit stdbit.h assert.h static_assert(__STDC_VERSION_STDBIT_H__ >= 202311L, "Compiler lacks stdbit.h")
test_code ld stdbit.h assert.h static_assert(__STDC_VERSION_STDBIT_H__ >= 202311L, "Compiler lacks stdbit.h") cc
test_ld cc
test_cc
BEGIN /tmp/ffconf.Y7bFUbHS/test.c
    1	#include <stdbit.h>
    2	#include <assert.h>
    3	int main(void) { static_assert(__STDC_VERSION_STDBIT_H__ >= 202311L, "Compiler lacks stdbit.h"); return 0; }
END /tmp/ffconf.Y7bFUbHS/test.c
ccache gcc -D_ISOC11_SOURCE -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -DWIN32_LEAN_AND_MEAN -U__STRICT_ANSI__ -D__USE_MINGW_ANSI_STDIO=1 -D__printf__=__gnu_printf__ -D_POSIX_C_SOURCE=200112 -D_XOPEN_SOURCE=600 -DPIC -D_FORTIFY_SOURCE=2 -fstack-protector-strong -pipe -D__USE_MINGW_ANSI_STDIO=1 -mthreads -Wno-int-conversion -std=c17 -fomit-frame-pointer -g -Wdeclaration-after-statement -Wall -Wdisabled-optimization -Wpointer-arith -Wredundant-decls -Wwrite-strings -Wtype-limits -Wundef -Wmissing-prototypes -Wstrict-prototypes -Wempty-body -Wno-parentheses -Wno-switch -Wno-format-zero-length -Wno-pointer-sign -Wno-unused-const-variable -Wno-bool-operation -Wno-char-subscripts -O3 -fno-math-errno -fno-signed-zeros -fno-tree-vectorize -Werror=format-security -Werror=implicit-function-declaration -Werror=missing-prototypes -Werror=return-type -Werror=vla -Wformat -fdiagnostics-color=auto -Wno-maybe-uninitialized -c -o /tmp/ffconf.Y7bFUbHS/test.o /tmp/ffconf.Y7bFUbHS/test.c
C:/mabszlib/msys64/tmp/ffconf.Y7bFUbHS/test.c:1:10: fatal error: stdbit.h: No such file or directory
    1 | #include <stdbit.h>
      |          ^~~~~~~~~~
compilation terminated.
ERROR: zlib requested but not found
```

L4cache commented:
I don't have problem with it but just so you know, sometimes we assume `1` for enable and `2` for disable, and larger for other choices, like most of other options...

woot000 commented:
> I don't have problem with it but just so you know, sometimes we assume `1` for enable and `2` for disable, and larger numbers for other choices, like most of other options...

I have the options set up like that because I'd rather not have the 'don't build' option set as `5`, in the case that I do manage to get zlib-rs working and add it in. This feature is also somewhat different from the others in that there is no 'disable' option for zlib, so I'd preferably have the 'don't build, use msys2 package` both more clearly separated from the rest and as an implied default (because using the zlib msys2 package is already the default for the suite)

L4cache commented:
I'd say the "keep old behavior" is the equivalent of "disable" in this situation.

L4cache commented:
I fine no problem builing ffmpeg with zlib-rs, what did you do?

woot000 commented:
> I find no problem builing ffmpeg with zlib-rs, what did you do?

I made a bit more progress with the zlib-rs build, but I'm still getting an error when linking it to ffmpeg:
```
LD	ffmpeg_g.exe
LD	ffprobe_g.exe
C:/mabszlib/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/14.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: libavformat/libavformat.a(http.o): in function `parse_content_encoding':
C:\mabszlib\build\ffmpeg-git\build-static-64bit/src/libavformat/http.c:888:(.text+0x1f0e): undefined reference to `zlibCompileFlags'
C:/mabszlib/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/14.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: libavformat/libavformat.a(http.o): in function `parse_content_encoding':
C:\mabszlib\build\ffmpeg-git\build-static-64bit/src/libavformat/http.c:888:(.text+0x1f0e): undefined reference to `zlibCompileFlags'
collect2.exe: error: ld returned 1 exit status
make: *** [/build/ffmpeg-git/Makefile:142: ffprobe_g.exe] Error 1
collect2.exe: error: ld returned 1 exit status
make: *** Waiting for unfinished jobs....
make: *** [/build/ffmpeg-git/Makefile:142: ffmpeg_g.exe] Error 1
LD	ffmpeg_g.exe
C:/mabszlib/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/14.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: libavformat/libavformat.a(http.o): in function `parse_content_encoding':
C:\mabszlib\build\ffmpeg-git\build-static-64bit/src/libavformat/http.c:888:(.text+0x1f0e): undefined reference to `zlibCompileFlags'
collect2.exe: error: ld returned 1 exit status
make: *** [/build/ffmpeg-git/Makefile:142: ffmpeg_g.exe] Error 1
```
For reference, this is what I have as the script currently:
```sh
if do_vcs "$SOURCE_REPO_ZLIBRS"; then
    do_uninstall "${_check[@]}" "${_zlib_uninstall[@]}" zlib_name_mangling.h
    cd_safe libz-rs-sys-cdylib
    # edit package metadata and library name to match zlib
    sed -e 's;libz_rs;libz;' -e 's;z_rs;z;' -i Cargo.toml
    PKG_CONFIG="$LOCALDESTDIR/bin/ab-pkg-config-static.bat" \
        log "rust.capi" cargo capi install \
        --release --jobs "$cpuCount" --prefix="$LOCALDESTDIR" \
        --destdir="$PWD/install-$bits"
    do_install "target/$CARCH-pc-windows-gnu$rust_target_suffix/release/libz.a" libz.a
    do_install "target/$CARCH-pc-windows-gnu$rust_target_suffix/release/libz.pc" zlib.pc
    do_checkIfExist
fi
```

L4cache commented:
You didn't install the headers?

woot000 commented:
> You didn't install the headers?

Installing the header (with `sed -e 's;libz_rs;libz;' -e 's;z_rs;z;' -e 's;enabled = false;enabled = true;' -i Cargo.toml` and `do_install "target/$CARCH-pc-windows-gnu$rust_target_suffix/release/z.h" zlib.h`) makes ffmpeg fail during configure:
```
ccache gcc -D_ISOC11_SOURCE -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -DWIN32_LEAN_AND_MEAN -U__STRICT_ANSI__ -D__USE_MINGW_ANSI_STDIO=1 -D__printf__=__gnu_printf__ -D_POSIX_C_SOURCE=200112 -D_XOPEN_SOURCE=600 -DPIC -D_FORTIFY_SOURCE=2 -fstack-protector-strong -pipe -D__USE_MINGW_ANSI_STDIO=1 -mthreads -Wno-int-conversion -std=c17 -fomit-frame-pointer -IC:/mabszlib/local64/include/opus -IC:/mabszlib/local64/include/opus -IC:/mabszlib/local64/include -IC:/mabszlib/local64/include -g -Wdeclaration-after-statement -Wall -Wdisabled-optimization -Wpointer-arith -Wredundant-decls -Wwrite-strings -Wtype-limits -Wundef -Wmissing-prototypes -Wstrict-prototypes -Wempty-body -Wno-parentheses -Wno-switch -Wno-format-zero-length -Wno-pointer-sign -Wno-unused-const-variable -Wno-bool-operation -Wno-char-subscripts -O3 -fno-math-errno -fno-signed-zeros -fno-tree-vectorize -Werror=format-security -Werror=implicit-function-declaration -Werror=missing-prototypes -Werror=return-type -Werror=vla -Wformat -c -o /tmp/ffconf.RM8Gf4Kq/test.o /tmp/ffconf.RM8Gf4Kq/test.c
check_cflags -fdiagnostics-color=auto
test_cflags -fdiagnostics-color=auto
test_cc -fdiagnostics-color=auto
BEGIN /tmp/ffconf.RM8Gf4Kq/test.c
    1	int x;
END /tmp/ffconf.RM8Gf4Kq/test.c
ccache gcc -D_ISOC11_SOURCE -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -DWIN32_LEAN_AND_MEAN -U__STRICT_ANSI__ -D__USE_MINGW_ANSI_STDIO=1 -D__printf__=__gnu_printf__ -D_POSIX_C_SOURCE=200112 -D_XOPEN_SOURCE=600 -DPIC -D_FORTIFY_SOURCE=2 -fstack-protector-strong -pipe -D__USE_MINGW_ANSI_STDIO=1 -mthreads -Wno-int-conversion -std=c17 -fomit-frame-pointer -IC:/mabszlib/local64/include/opus -IC:/mabszlib/local64/include/opus -IC:/mabszlib/local64/include -IC:/mabszlib/local64/include -g -Wdeclaration-after-statement -Wall -Wdisabled-optimization -Wpointer-arith -Wredundant-decls -Wwrite-strings -Wtype-limits -Wundef -Wmissing-prototypes -Wstrict-prototypes -Wempty-body -Wno-parentheses -Wno-switch -Wno-format-zero-length -Wno-pointer-sign -Wno-unused-const-variable -Wno-bool-operation -Wno-char-subscripts -O3 -fno-math-errno -fno-signed-zeros -fno-tree-vectorize -Werror=format-security -Werror=implicit-function-declaration -Werror=missing-prototypes -Werror=return-type -Werror=vla -Wformat -fdiagnostics-color=auto -c -o /tmp/ffconf.RM8Gf4Kq/test.o /tmp/ffconf.RM8Gf4Kq/test.c
test_cflags -Wmaybe-uninitialized
test_cc -Wmaybe-uninitialized
BEGIN /tmp/ffconf.RM8Gf4Kq/test.c
    1	int x;
END /tmp/ffconf.RM8Gf4Kq/test.c
ccache gcc -D_ISOC11_SOURCE -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -DWIN32_LEAN_AND_MEAN -U__STRICT_ANSI__ -D__USE_MINGW_ANSI_STDIO=1 -D__printf__=__gnu_printf__ -D_POSIX_C_SOURCE=200112 -D_XOPEN_SOURCE=600 -DPIC -D_FORTIFY_SOURCE=2 -fstack-protector-strong -pipe -D__USE_MINGW_ANSI_STDIO=1 -mthreads -Wno-int-conversion -std=c17 -fomit-frame-pointer -IC:/mabszlib/local64/include/opus -IC:/mabszlib/local64/include/opus -IC:/mabszlib/local64/include -IC:/mabszlib/local64/include -g -Wdeclaration-after-statement -Wall -Wdisabled-optimization -Wpointer-arith -Wredundant-decls -Wwrite-strings -Wtype-limits -Wundef -Wmissing-prototypes -Wstrict-prototypes -Wempty-body -Wno-parentheses -Wno-switch -Wno-format-zero-length -Wno-pointer-sign -Wno-unused-const-variable -Wno-bool-operation -Wno-char-subscripts -O3 -fno-math-errno -fno-signed-zeros -fno-tree-vectorize -Werror=format-security -Werror=implicit-function-declaration -Werror=missing-prototypes -Werror=return-type -Werror=vla -Wformat -fdiagnostics-color=auto -Wmaybe-uninitialized -c -o /tmp/ffconf.RM8Gf4Kq/test.o /tmp/ffconf.RM8Gf4Kq/test.c
ar rcD /tmp/ffconf.RM8Gf4Kq/test.a @/dev/null
check_builtin stdbit stdbit.h assert.h static_assert(__STDC_VERSION_STDBIT_H__ >= 202311L, "Compiler lacks stdbit.h")
test_code ld stdbit.h assert.h static_assert(__STDC_VERSION_STDBIT_H__ >= 202311L, "Compiler lacks stdbit.h") cc
test_ld cc
test_cc
BEGIN /tmp/ffconf.RM8Gf4Kq/test.c
    1	#include <stdbit.h>
    2	#include <assert.h>
    3	int main(void) { static_assert(__STDC_VERSION_STDBIT_H__ >= 202311L, "Compiler lacks stdbit.h"); return 0; }
END /tmp/ffconf.RM8Gf4Kq/test.c
ccache gcc -D_ISOC11_SOURCE -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -DWIN32_LEAN_AND_MEAN -U__STRICT_ANSI__ -D__USE_MINGW_ANSI_STDIO=1 -D__printf__=__gnu_printf__ -D_POSIX_C_SOURCE=200112 -D_XOPEN_SOURCE=600 -DPIC -D_FORTIFY_SOURCE=2 -fstack-protector-strong -pipe -D__USE_MINGW_ANSI_STDIO=1 -mthreads -Wno-int-conversion -std=c17 -fomit-frame-pointer -IC:/mabszlib/local64/include/opus -IC:/mabszlib/local64/include/opus -IC:/mabszlib/local64/include -IC:/mabszlib/local64/include -g -Wdeclaration-after-statement -Wall -Wdisabled-optimization -Wpointer-arith -Wredundant-decls -Wwrite-strings -Wtype-limits -Wundef -Wmissing-prototypes -Wstrict-prototypes -Wempty-body -Wno-parentheses -Wno-switch -Wno-format-zero-length -Wno-pointer-sign -Wno-unused-const-variable -Wno-bool-operation -Wno-char-subscripts -O3 -fno-math-errno -fno-signed-zeros -fno-tree-vectorize -Werror=format-security -Werror=implicit-function-declaration -Werror=missing-prototypes -Werror=return-type -Werror=vla -Wformat -fdiagnostics-color=auto -Wno-maybe-uninitialized -c -o /tmp/ffconf.RM8Gf4Kq/test.o /tmp/ffconf.RM8Gf4Kq/test.c
C:/mabszlib/msys64/tmp/ffconf.RM8Gf4Kq/test.c:1:10: fatal error: stdbit.h: No such file or directory
    1 | #include <stdbit.h>
      |          ^~~~~~~~~~
compilation terminated.
ERROR: zlib requested but not found
```

L4cache commented:
I just ran the `cargo build --release` in zlib-rs top-level folder, then I can't find the way to install it, so I manually searched and copied the files in `zlib-rs\target\release\build\libz-sys-e293eea3b4ce24bc\out` to the install destination, and FFmpeg build "just works".
I have no idea about how this rust stuff is supposed to be used, but it worked (shrug).

woot000 commented:
> I just ran the `cargo build --release` in zlib-rs top-level folder, then I can't find the way to install it, so I manually searched for the `.a` archive and header, and then copied the files in `zlib-rs\target\release\build\libz-sys-e293eea3b4ce24bc\out` to the install destination, and FFmpeg build "just works". I have no idea about how this rust stuff is supposed to be used, but it worked (shrug).

Those files in `libz-sys-e293eea3b4ce24bc\out` that you describe are apart of the `libz-sys` crate and not zlib-rs's drop-in replacement files. Based on the main project's `Cargo.toml` file (line 36: `libz-sys = { version = "1.1.21", default-features = false, features = ["zlib-ng"] } # use libz-ng in libz compat mode`), it appears to just be zlib-ng compiled with zlib compatible API.

After doing a little research, I found that the reason for the earlier ``undefined reference to `zlibCompileFlags'`` error is because that function (along with all the public inflate and gz functions) are [currently not implemented in zlib-rs](https://github.com/trifectatechfoundation/zlib-rs/issues/49).

With this information, I think I'll add zlib-rs at a later date, once at least `zlibCompileFlags` and the public inflate functions are implemented. I think at this current state, this pull request is good to merge.

L4cache commented:
> it appears to just be zlib-ng compiled with zlib compatible API.

Do they pull zlib-ng code from web? Thre repo itself does contain zlib-ng code.

woot000 commented:
> Do they pull zlib-ng code from web? Thre repo itself doesn't contain zlib-ng code.

Yes, the `zlib-sys` crate (built as a dependency for zlib-rs) lists [zlib-ng as a submodule](https://github.com/rust-lang/libz-sys/blob/15c7dba20288dea32a91b0c2d86fdf21c25d0f63/.gitmodules#L6)

woot000 commented:
I've added zlib-rs since they've recently implemented [zlibCompileFlags](https://github.com/trifectatechfoundation/zlib-rs/commit/b70b7abeeea73e08cd07dbdff206edb1e037e19d) a few days ago, and ffmpeg links with it just fine in both 32-bit and 64-bit compilation.
I also added the `-DUSE_ZLIB_RABIN_KARP_HASH=ON` option for zlib-chromium, which writes its output as bitstream compatible with zlib (not needed for linking to other programs, it's just nice to have a faster zlib alternative that doesn't produce larger .zips) and rearranged the options so that "disable" is at the bottom, similar to how I've set up av1an and gifski's options.

@1480c1 Unless you find some issue or oversight I missed, I think this PR is ready to be merged now

1480c1 commented:
Thanks everyone

L4cache commented:
> the `-DUSE_ZLIB_RABIN_KARP_HASH=ON` option for zlib-chromium, which writes its output as bitstream compatible with zlib (not needed for linking to other programs, it's just nice to have a faster zlib alternative that doesn't produce larger .zips)

I tested it a bit and found it basically performs the same as the OG zlib when this additional compatibility is enabled.
